### ISSUE TITLE
sql: dropping a column does not drop partial unique constraints referencing that column

### ISSUE THREAD

**Issue by mgartner** [cockroachdb/cockroach#97370]
When dropping a column, a `UNIQUE WITHOUT INDEX` constraint with a partial predicate referencing the dropped column is not dropped itself. This causes errors. I suspect this affect RBR tables that implicitly create `UNQIUE WITHOUT INDEX` constraints.

```
defaultdb> SET experimental_enable_unique_without_index_constraints=on;
SET

defaultdb> SET sql_safe_updates=false;
SET

defaultdb> CREATE TABLE t (k INT PRIMARY KEY, a INT, b INT, UNIQUE WITHOUT INDEX (b) WHERE a > 0);
CREATE TABLE

defaultdb> SHOW CREATE TABLE t;
  table_name |                          create_statement
-------------+----------------------------------------------------------------------
  t          | CREATE TABLE public.t (
             |     k INT8 NOT NULL,
             |     a INT8 NULL,
             |     b INT8 NULL,
             |     CONSTRAINT t_pkey PRIMARY KEY (k ASC),
             |     CONSTRAINT unique_b UNIQUE WITHOUT INDEX (b) WHERE a > 0:::INT8
             | )
(1 row)


defaultdb> ALTER TABLE t DROP COLUMN a;
ERROR: error executing StatementPhase stage 1 of 1 with 9 MutationType ops: relation "t" (106): partial unique without index constraint "unique_b" refers to unknown columns in predicate: a > 0:::INT8
```

Jira issue: CRDB-24660

**Comment by ajwerner** [comment]
This is important and we'll fix it.

**Comment by Xiang-Gu** [comment]
Update: We recently encountered [an issue](https://github.com/cockroachdb/cockroach/issues/96924) with dropping a column referenced in a partial index's predicate (or a partial unique without index constraint's predicate). As a temporary fix, we disallowed this behavior and instead hint user to drop the partial index (or the constraint) first before dropping the table. We'd like to lift this limitation in the future and that is tracked in https://github.com/cockroachdb/cockroach/issues/97813. I'm going to close this issue in favor of that issue.
